contracts/gnosis_safe_modules/DelayedRecoveryModule.sol

```diff
@@ -0,0 +1,627 @@
+     * @param _recoveryController An address that relays signed requests of
+     *                            different types.
+     * @param _recoveryBlockDelay A required number of blocks to pass to
+     *                            be able to execute a recovery request.
+     */
+    function setup(
+        address _recoveryOwner,
+        address _recoveryController,
+        uint256 _recoveryBlockDelay
+    )
+        external
+    {
+        // This check assures that the setup function is called only once.
+        require(
+            domainSeparator == bytes32(0),
+            "Domain separator was already set."
+        );
+
+        require(
+            _recoveryOwner != address(0),
+            "Recovery owner's address is null."
+        );
+
+        require(
+            _recoveryController != address(0),
+            "Recovery controller's address is null."
+        );
+
+        domainSeparator = keccak256(
+            abi.encode(
+                DOMAIN_SEPARATOR_TYPEHASH,
+                this
+            )
+        );
+
+        manager = GnosisSafeModuleManagerInterface(msg.sender);
+
+        recoveryOwner = _recoveryOwner;
+
+        recoveryController = _recoveryController;
+
+        recoveryBlockDelay = _recoveryBlockDelay;
+    }
+
+    /**
+     * @notice Initiates a recovery procedural.
+     *
+     * @dev Function requires:
+     *          - Only the recovery controller can call.
+     *          - There is no active recovery procedural.
+     *          - Recovery owner has signed message.
+     *
+     * @param _prevOwner Owner that pointed to the owner to be replaced in the
+     *                   linked list.
+     * @param _oldOwner Owner address to replace.
+     * @param _newOwner New owner address.
+     */
+    function initiateRecovery(
+        address _prevOwner,
+        address _oldOwner,
+        address _newOwner,
+        bytes32 _r,
+        bytes32 _s,
+        uint8 _v
+    )
+        external
+        onlyRecoveryController
+        noActiveRecovery
+    {
+        bytes32 recoveryHash = hashInitiateRecovery(
+            _prevOwner,
+            _oldOwner,
+            _newOwner
+        );
+
+        verify(recoveryHash, _r, _s, _v);
+
+        activeRecoveryInfo = RecoveryInfo({
+            prevOwner: _prevOwner,
+            oldOwner: _oldOwner,
+            newOwner: _newOwner,
+            initiationBlockHeight: block.number,
+            initiated: true
+        });
+
+        emit RecoveryInitiated(
+            _prevOwner,
+            _oldOwner,
+            _newOwner
+        );
+    }
+
+    /**
+     * @notice Executes the initiated recovery.
+     *
+     * @dev Function requires:
+     *          - Only recovery controller can call.
+     *          - There is an initiated recovery with the same tuple of
+     *            addresses (prevOwner, oldOwner, newOwner).
+     *          - Recovery owner has signed execution message.
```

Comment:
as previously commented, recovery owner does not need to sign for execution

Reply:
Done.